Source/InvocationCollection.cs

```diff
@@ -78,46 +78,101 @@ public int Count
 
 		public void Add(Invocation invocation)
 		{
-			lock (this.invocations)
+			lock (this.invocationsLock)
 			{
-				this.invocations.Add(invocation);
+				if (this.count == this.capacity)
+				{
+					var targetCapacity = this.capacity == 0 ? 4 : (this.capacity * 2);
+					Array.Resize(ref this.invocations, targetCapacity);
+					this.capacity = targetCapacity;
+				}
+
+				this.invocations[this.count] = invocation;
+				this.count++;
 			}
 		}
 
 		public bool Any()
 		{
-			return this.invocations.Count > 0;
+			lock (this.invocationsLock)
+			{
+				return this.count > 0;
+			}
 		}
 
 		public void Clear()
 		{
-			lock (this.invocations)
+			lock (this.invocationsLock)
 			{
-				this.invocations.Clear();
+				// Replace the collection so readers with a reference to the old collection aren't interrupted
+				this.invocations = null;
+				this.count = 0;
+				this.capacity = 0;
 			}
 		}
 
 		public Invocation[] ToArray()
 		{
-			lock (this.invocations)
+			lock (this.invocationsLock)
 			{
-				return this.invocations.ToArray();
+				if (this.count == 0)
+				{
+					return new Invocation[0];
+				}
+
+				var result = new Invocation[this.count];
+
+				Array.Copy(this.invocations, result, this.count);
+
+				return result;
 			}
 		}
 
 		public Invocation[] ToArray(Func<Invocation, bool> predicate)
 		{
-			lock (this.invocations)
+			lock (this.invocationsLock)
 			{
-				return this.invocations.Where(predicate).ToArray();
+				if (this.count == 0)
+				{
+					return new Invocation[0];
+				}
+
+				var result = new Invocation[this.count];
+				var resultIndex = 0;
+				var resultSize = 0;
+
+				for (var i = 0; i < this.count; i++)
+				{
+					var invocation = this.invocations[i];
+					if (predicate(invocation))
+					{
+						result[resultIndex] = invocation;
+						resultIndex++;
+						resultSize++;
+					}
+				}
+
+				Array.Resize(ref result, resultSize);
```

Comment:
Haha, I know I previously said that I tend to avoid LINQ queries in low-level places such as this class... but please feel free to just use a `List<>` with an initial capacity of `this.count`, then returning `.ToArray()` in the end. It's a common pattern that would save you some lines of code at only very little additional overhead.